KAFKA-9450: Decouple flushing state from commiting

guozhangwang opened the pull request:
In Kafka Streams the source-of-truth of a state store is in its changelog, therefore when committing a state store we only need to make sure its changelog records are all flushed and committed, but we do not actually need to make sure that the materialized state have to be flushed and persisted since they can always be restored from changelog when necessary.

On the other hand, flushing a state store too frequently may have side effects, e.g. rocksDB flushing would gets the memtable into an L0 sstable, leaving many small L0 files to be compacted later, which introduces larger overhead.

Therefore this PR decouples flushing from committing, such that we do not always flush the state store upon committing, but only when sufficient data has been written since last time flushed. The checkpoint file would then also be overwritten only along with flushing the state store indicating its current known snapshot. This is okay since: a) if EOS is not enabled, then it is fine if the local persisted state is actually ahead of the checkpoint, b) if EOS is enabled, then we would never write a checkpoint file until close.

Here's a more detailed change list of this PR:

1. Do not always flush state stores when calling pre-commit; move `stateMgr.flush` into post-commit to couple together with checkpointing. 

2. In post-commit, we checkpoint when: a) The state store's snapshot has progressed much further compared to the previous checkpoint, b) When the task is being closed (but as a minor optimization, we would avoid checkpointing on closing if it is exactly the same as the previous one).

3. There are some tricky obstacles that I'd have to work around in a bit hacky way: for cache / suppression buffer, we still need to flush them in pre-commit to make sure all records sent via producers, while the underlying state store should not be flushed. I've decided to introduce a new API in CachingStateStore to be triggered in pre-commit.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

guozhangwang committed:
first commit

guozhangwang committed:
major fixes

guozhangwang committed:
flush cache before commit

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

guozhangwang committed:
suppression buffer needs to be flushed too

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:
```diff
@@ -38,13 +41,39 @@
+
+        // when enforcing checkpoint is required, we should overwrite the checkpoint if it is different from the old one;
+        // otherwise, we only overwrite the checkpoint if it is largely different from the old one
+        return enforceCheckpoint ? totalOffsetDelta > 0 : totalOffsetDelta > OFFSET_DELTA_THRESHOLD_FOR_CHECKPOINT;
```
I'm a bit torn about this optimization to avoid double checkpointing, because on the other hand, even if we have not made any progress since loaded the checkpoint, we'd still need to make a checkpoint upon closing if we have never made one before -- and I use emptyMap as an indicator for that.

Given that upon suspending we are now less likely to checkpoint, the chance that we would double checkpointing (when transiting to suspended, and when transiting to closed) is smaller, and hence I'm thinking maybe I'd just remove this optimization to make the logic a bit simpler. LMK WDYT.

ableegoldman replied:
Is this out of date? It seems like we now never checkpoint during suspension so we don't have to bother with this optimization

guozhangwang replied:
Sorry I meant the "commit" during suspension: when we suspend a task we would commit the offsets and in `postCommit` we may write the checkpoint file, and then when we close the suspended task we may commit again (note the `prepareCommit` call should return empty offsets since there's nothing new to commit, but in `postCommit` we would checkpoint the file again, i.e. double writing checkpoint.

So what I did here is to not enforce checkpointing in `postCommit` during suspension, and only enforce checkpointing in `postCommit` during closing.

ableegoldman replied:
I'm still not following -- I thought we no longer commit during close at all, we just commit when suspending? For active tasks we call pre/postCommit and suspend in `handleRevocation` and then just close them in `handleAssignment`. For standbys we do all those things in `handleAssignment`, but still only once

ableegoldman replied:
Oh, I see, we now do the checkpoint in `handleAssignment` as well. Why? It seems like we're just bringing back the problem of potentially double-checkpointing that had been solved?

ableegoldman replied:
Can you elaborate on why not enforce the checkpoint during suspension in `handleRevocation` (for the tasks to be closed)? It seems like we just re-introduce this problem of potentially double checkpointing and I'm not sure I understand why we need to checkpoint only in `handleAssignment`. Is this an optimization or a correctness issue or a code cleanness thing?

guozhangwang replied:
I think we do not need to enforce checkpoint during suspension but only need to do that during closure / recycling; if a suspended task is resumed then we do not need to write checkpoint in between.

But admittedly moving forward most suspended tasks would be closed or recycled :slightly_smiling_face: So I can change that back.

mjsax replied:
I would expect that writing the checkpoint file should be cheap anyway (it's just some metadata, how bad can it be?) -- or are you worried about the blocking flush to the file system? Thus, I don't consider double-checkpointing as a real issue? Might be a case of pre-mature optimization? It might simplify our code if we just blindly write the checkpoint data even if it did not change? -- But I am fine either way.

guozhangwang replied:
In the current code we actually call `postCommit(true)` both during suspension and closure. It's just that in `checkpointNeeded` we would only return true with `enforce` if it has changed --- personally I think this is a general optimization worth applying.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -542,13 +530,22 @@ public void closeCleanAndRecycleState() {
+     *                          or flushing state store get IO errors; such error should cause the thread to die
+     */
+    @Override
+    protected void maybeWriteCheckpoint(final boolean enforceCheckpoint) {
```
I decided to extract out the update of the changelog offsets from actually writing the offsets since even if we do not want to write the file, we still need to update the offsets.

The reason I did not yet remove the parameter from `checkpoint` is that global-task is still using it. I plan to remove it when consolidating the global task.

ableegoldman replied:
Was there a reason to not just add `#updateChangelogOffsets` to the `StateManager` interface and remove the checkpointable offsets argument from `#checkpoint`?

guozhangwang replied:
Let me try and see if that works; will get back to you.

guozhangwang reviewed (commented):
ping @ableegoldman @vvcephei for reviews.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -478,9 +479,11 @@ void handleRevocation(final Collection<TopicPartition> revokedPartitions) {
         for (final Task task : revokedTasks) {
             try {
-                task.postCommit();
+                task.postCommit(false);
```
Inspired by https://github.com/apache/kafka/pull/8962/files#diff-f0037ae2fd44bdd0d84f3ef37b43bc05R277, I think we should actually enforce checkpoint (i.e. set to `true`) when suspending since the suspended task may not be closed but recycled. cc @ableegoldman 

I will push a new commit changing to true.

guozhangwang replied:
Actually after rebased on latest trunk I think upon suspending we can still not enforcing the checkpointing, but only enforce upon closing / recycling in handleAssignment.

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

guozhangwang committed:
rebased on trunk

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:
```diff
@@ -38,13 +41,39 @@
  */
 final class StateManagerUtil {
     static final String CHECKPOINT_FILE_NAME = ".checkpoint";
+    static final long OFFSET_DELTA_THRESHOLD_FOR_CHECKPOINT = 10_000L;
```
This is just a placeholder for making less frequent flushing than commits, currently I'm making it num.records based, but it's open for better proposal to decide when flushing should be executed.

ableegoldman replied:
It would be awesome if we could do it based on the size of the pending data and the configured memtable size. Not sure if that's really feasible, just throwing it out there

guozhangwang replied:
You mean when the accumulated pending data is larger than the memtable, we should checkpoint?

ableegoldman replied:
Something like that. We know rocksdb will render the memtable immutable once it reaches the configured memtable size, after that it will flush once the number of immutable memtables reaches the configured value. Probably makes sense to align our checkpoint/flushing to the configured rocksdb flushing.

Would be cool if we could piggy-back on the rocksdb options and avoid a new config in Streams altogether, but obviously not everyone uses rocksdb

guozhangwang replied:
Yeah my major concern is to tie the flushing policy with rocksdb -- although it is the default persistent stores now, we should avoid tying with a specific type of stores.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -454,6 +456,41 @@ public void flush() {
+
+                try {
+                    // buffer should be flushed to send all records to changelog
+                    if (store instanceof TimeOrderedKeyValueBuffer) {
```
Admit this is kinda hacky, but I'd have to do this for cached store and suppression buffer. Moving forward I think the first can be removed when we decouple caching with emitting, but for suppression buffer maybe we can have a more general way to fix it, for example maybe we could just have changelogger to always buffer itself so that suppression buffers do not need to buffer itself to changelogger. cc @vvcephei

ableegoldman replied:
WDYT about adding a generic `FlushingRequiredStore` marker interface that all caching state stores and the suppression buffer would both implement. It seems weird to handle them separately.

We could even make this public and allow user custom state stores to implement this, but that might be opening a can of worms we will greatly regret 😉

guozhangwang replied:
My plan is actually to remove the `flushCache` once we decoupled caching with emitting (see the TODO comment on the caller).

ableegoldman replied:
So you're saying we'd still need to flush the suppression buffer but not the cache once we decouple caching from emitting? Or that we can remove this `flushCache` method altogether once that is done? Or that it will still do some flushing, but will not resemble the current `flushCache` method at all

guozhangwang replied:
I'm thinking we can remove the whole `flushCache` method.

ableegoldman replied:
Is there a ticket for that?

mjsax replied:
I agree. IMHO, we should get rid of KTable/store cache all together and only have a "changelog-writer-cache" that regular stores and suppress() can use to reduce the write load on the changelog topic. For downstream rate control, users should use suppress() (and we might want to try to unify suppress() and the upstream store somehow eventually to avoid the current redundancy)

guozhangwang replied:
I'm going to create the ticket of that very soon.

guozhangwang reviewed (commented):
I've rebased this PR on trunk and it is officially open for reviews. @ableegoldman @vvcephei

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -243,18 +242,24 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
 
         for (final Task task : tasksToClose) {
             try {
-                if (task.isActive()) {
```
As we discussed in the other PR, I'm removing the logic for TM to check on task state and replaced it with checking that there should be nothing to commit (since suspend / pre-commit / post-commit are not all idempotent).

mjsax replied:
What other PR?

guozhangwang replied:
This is the PR from @ableegoldman https://github.com/apache/kafka/pull/8962

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -454,6 +456,41 @@ public void flush() {
+                        if (exception instanceof StreamsException)
+                            firstException = exception;
+                        else
+                            firstException = new ProcessorStateException(
```
nit: can you use braces here?

guozhangwang replied:
Ack

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -520,20 +557,27 @@ void transitionTaskType(final TaskType newType, final LogContext logContext) {
+        // first update each state store's current offset, then checkpoint
+        // those stores that are only logged and persistent to the checkpoint file
+        // TODO: we still need to keep it as part of the checkpoint for global tasks; this could be
+        //       removed though when we consolidate global tasks / state managers into this one
```
>TODO: we still need to keep it as part of the checkpoint for global tasks

Took me a while to realize that "it" refers to the argument here -- can you clarify the comment?

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -120,14 +120,12 @@ public void suspend() {
         switch (state()) {
             case CREATED:
                 log.info("Suspended created");
-                checkpointNeededForSuspended = false;
```
Happy to see this go 🙂

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -49,6 +61,31 @@
+            // since there's no written offsets we can checkpoint with empty map,
+            // and the state's current offset would be used to checkpoint
+            stateMgr.flush();
+            stateMgr.checkpoint(Collections.emptyMap());
```
Should we add an assertion to `ProcessorStateManager#checkpoint` that the passed in offsets are always empty?

guozhangwang replied:
The passed in parameters are not always empty unfortunately since global tasks would still pass in non-empty map.

ableegoldman replied:
But `ProcessorStateManager` doesn't handle global tasks

guozhangwang replied:
I see what did you mean now. Will do.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -243,18 +242,24 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                final Map<TopicPartition, OffsetAndMetadata> offsets = task.prepareCommit();
+
+                if (!offsets.isEmpty()) {
+                    log.error("Task {} should has been committed prior to attempting to close, but it reports non-empty offsets {} to commit",
```
Hm. What if we hit a TaskMigratedException during `handleRevocation`? We would never finish committing them so `commitNeeded` would still return true and `prepareCommit` would return non-empty offsets right?

It's kind of a bummer that we can't enforce that the task was committed. What we really need to do is enforce that we _attempted_ to commit the task -- regardless of whether or not it was successful. If the commit failed we know that either it was fatal or it was due to TaskMigrated, in which case the task will have to be closed as dirty anyways.
This might be beyond the scope of this PR, but just to throw out one hacky idea we could add a `commitSuccessful` parameter to `postCommit` and then always invoke that after a commit so that `commitNeeded` is set to false. (If `commitSuccessful` is false we just skip everything else in `postCommit`)

guozhangwang replied:
I thought if we hit a TaskMigrated during `handleRevocation` then it would be thrown all the way from consumer and cause the thread to handle it via `handleLostAll` and hence `handleAssignment` should not be called. 

After checking the source code I think not callers are throwing the exception directly and may wrap it as a KafkaException, but even in that case we would just crash the client, right?

ableegoldman replied:
I thought we would catch and save the first exception thrown by a rebalance listener callback, and then rethrow after all rebalance callbacks have been invoked? In this case that would mean `handleAssignment` would still get called, and then we would throw an IllegalStateException and bail on the rest of `handleAssignment` for no reason.

The IllegalStateException itself is not the problem, since only the first exception (the TaskMigrated) would ultimately be thrown up to `poll`. But we should still go through the rest of `handleAssignment` in order to properly clean up the active tasks and manage the standbys (since we don't need to close standbys in case of TaskMigrated)

guozhangwang replied:
I've reordered in handleRevocation so that commit is called before suspension; also even if commit failed with TaskMigrated, we would still try to continue completing the suspension so in handleAssignment we should have all active tasks suspended or should be closed dirty.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -466,15 +458,8 @@ public void postCommit() {
                  */
                 partitionGroup.clear();
```
This is also maybe beyond the scope of this PR, but it seems like there's no reason to do things like this anymore. Specifically, today we enforce the order `suspend` -> `pre/postCommit` -> `close` where `suspend` only closes the topology and we only use the `SUSPENDED` state to enforce that we suspended before closing. Why not swap the order so that we `pre/postCommit` -> `suspend` -> `close` and then we can move this call from `postCommit` to `suspend` where it makes more sense. Thoughts?

guozhangwang replied:
Hmm, what's the original reason that we must `suspend` before `commit`?

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -93,8 +93,8 @@ public boolean isActive() {
         if (state() == State.CREATED) {
             StateManagerUtil.registerStateStores(log, logPrefix, topology, stateMgr, stateDirectory, processorContext);
-
             // initialize the snapshot with the current offsets as we don't need to commit then until they change
```
Can you move this comment down a line? We should avoid confusion since we actually don't initialize the (flush) snapshot with the current offsets, just the (commit) snapshot.

To be honest, it's already pretty confusing that we initialize the two snapshots differently. Maybe you could add a quick sentence explaining why for our own future reference

guozhangwang replied:
We maintain two snapshots in standby because its `commitNeeded` is overridden such that we only commit if the offsets has changed since last time committed, whereas in active we have the flag itself to indicate so.

I can add the above explanation for `offsetSnapshotSinceLastCommit`, would that be clearer?

ableegoldman replied:
Sounds good. I think the most important thing is to not have the comment that says `initialize the snapshot with the current offsets...` right above the line that initializes the snapshot as empty 🙁

guozhangwang replied:
Ack

ableegoldman reviewed (commented):
Left some initial thoughts @guozhangwang. I haven't gone through the testing code yet but I wanted to leave the first set of review comments since it's been so long already. Thanks for being patient! 🙂

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

guozhangwang committed:
github comments

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

guozhangwang committed:
re-order commit and suspend

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

guozhangwang committed:
use list not set to preserve ordering

guozhangwang commented:
@ableegoldman @vvcephei I've thought about controlling the frequency of flushing, and I've decided to expose another public config to let users specify. But I'd defer that to another PR. I'd like to do the following:

1) Merge this PR based on a hard-coded num.records difference.
2) Run benchmarks to understand its implications on stateful applications with / wo EOS.
3) Then file a KIP that introduce a new config value, also potentially change the default commit interval with EOS.

guozhangwang committed:
enforce checkpoint during suspension

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +313,96 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                //    and their changelog positions should not change at all postCommit would not write the checkpoint again.
+                // 2) for standby tasks prepareCommit should always return empty, and then in postCommit we would probably
+                //    write the checkpoint file.
+                final Map<TopicPartition, OffsetAndMetadata> offsets = task.prepareCommit();
```
It's still making me a bit uncomfortable that we call `prepare/postCommit` during `handleRevocation`. and `handleAssignment`. Maybe I'm being paranoid but experience has shown it's been difficult for us to keep track of which methods need to be called when, and in what order.
It seems like, now that we've decoupled flushing from committing, the only reason for calling pre/postCommit in `handleRevocation` is so that the record collector is flushed before committing offsets. So what if we extract the record collector flushing out into a separate StreamTask method that is only ever called in `TaskManager#commitOffsetsOrTransaction`? I haven't thought this all the way through but it just seems like we may as well go all the way in decoupling flushing from committing and split them out into separate methods. Maybe `preCommit` and `postCommit` have become relics of the past

ableegoldman replied:
I know that's kind of another large change, so feel free to tell me to drop it 🙂  Or one of us can consider as followup work.

ableegoldman replied:
Alternatively, now that we enforce checkpoint during suspension, we could just remove the `pre/postCommit` for active tasks in `handleAssignment`. It just seems nice to be able to assert that we never call `pre/postCommit` after a task is suspended

guozhangwang replied:
I'm leaning towards the second option here: in `handleAssignment` we actually do not commit at all, but only use 1) `preCommit` to validate certain states, and use 2) `postCommit` and expecting it to be a no-op actually. Let me see if I can get rid of those two in `handleAssignment`.

guozhangwang replied:
I tried some ways and ended-up with explicitly specifying suspend / postCommit for `standby` tasks only, and use `prepareCommit` to check if the previous revocation has failed or not. Personally I'm happy with the current workflow now.

vvcephei commented:
Sounds good, @guozhangwang . Another option is to make an internal config so that we can parameterize the benchmarks and get a more thorough understanding of the impact of this change.

Of course, we can easily transition from the hard-coded value to an internal config in a small follow-on PR.

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

ableegoldman commented:
Kicked off system tests https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4081

guozhangwang committed:
minor tweaks

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -461,6 +463,42 @@ public void flush() {
+                        store.flush();
+                    } else if (store instanceof CachedStateStore) {
+                        ((CachedStateStore) store).flushCache();
+                    }
```
Seems like there's the missing possibility that it's not TimeOrdered or Cached. Should we log a different message than "Flushed cache or buffer" in that case, to indicate we _didn't_ flush it?

guozhangwang replied:
For stores that's not time-ordered or cached, we should not flush them indeed. In fact moving forward I think we would not flush cache store anyways since they will be removed. I.e. generally speaking we should not `flush cache` always. In that sense the log4j entry looks reasonable to me?

ableegoldman reviewed (approved):
Thanks @guozhangwang , this looks reasonable to me. I had a typo in the branch name on the system test run so I just kicked off a new set. Assuming that's all green, LGTM

We do have an integration test that verifies EOS is not violated on a unclean shutdown, right? That seems sufficient for covering this case, just want to make sure everything behaves as expected

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:
```diff
@@ -38,13 +41,39 @@
+        // we can checkpoint if the the difference between the current and the previous snapshot is large enough
+        long totalOffsetDelta = 0L;
+        for (final Map.Entry<TopicPartition, Long> entry : newOffsetSnapshot.entrySet()) {
+            totalOffsetDelta += Math.abs(oldOffsetSnapshot.getOrDefault(entry.getKey(), 0L) - entry.getValue());
```
What's the rationale for subtracting the larger value from the smaller one and then taking the absolute value?

guozhangwang replied:
Was trying to make sure we do not get an NPE but I think that's not necessary.. will change.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -267,80 +266,26 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
 
-        for (final Task task : dirtyTasks) {
+        // for tasks that cannot be cleanly closed or recycled, close them dirty
+        for (final Task task : tasksToCloseDirty) {
```
It seems a little odd to have `handleCloseAndRecycle` not do this but just update the taskToCloseDirty list, since it handles everything else.

guozhangwang replied:
That's a good point, I will update.

vvcephei reviewed (commented):
Hey @guozhangwang , I'm about halfway through, and out of time today.

This is looking good overall, just a couple of minor thoughts.

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

guozhangwang committed:
github comment, plus a fix for system test

guozhangwang commented:
cc @mjsax please take a final look.

guozhangwang committed:
fix unit tests

guozhangwang committed:
improved eos test driver

guozhangwang committed:
minor fix

guozhangwang committed:
tweak unit tests

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

guozhangwang committed:
commit first then suspend in closing

guozhangwang committed:
first commit then suspend in closing

guozhangwang committed:
remove debugging info

guozhangwang commented:
System test green: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4093/

Ready for a final review and merge cc @mjsax

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -49,6 +61,30 @@
     }
 
+    protected void initializeCheckpoint() {
+        // we will delete the local checkpoint file after registering the state stores and loading them into the
```
Why do we actually delete the checkpoint file after registering? For non-eos, it seems we only need to delete the checkopint file if we wipe out the whole store?

Or is it a "simplification" do not distinguish between eos/non-eos for this case?

guozhangwang replied:
Yes it is a simplification: we are going to write the checkpoint file after initialization and on the first commit we would write checkpoint file again, so that we only have a short window of risk.

mjsax replied:
Independent of this PR, might it be worth to change this and close this window of risk?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java:
```diff
@@ -390,9 +399,12 @@ private static void verifyReceivedAllRecords(final Map<TopicPartition, List<Cons
                 }
             }
+
+            if (exception != null)
```
Nit: add `{ }`

guozhangwang replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java:
```diff
@@ -379,9 +379,18 @@ private static void verifyReceivedAllRecords(final Map<TopicPartition, List<Cons
+            final Iterator<ConsumerRecord<byte[], byte[]>> expectedRecord = expectedRecordsForPartition.iterator();
+            RuntimeException exception = null;
+            for (final ConsumerRecord<byte[], byte[]> receivedRecord : receivedRecordsForPartition) {
+                if (!expectedRecord.hasNext())
```
Nit: add `{ }`

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java:
```diff
@@ -379,9 +379,18 @@ private static void verifyReceivedAllRecords(final Map<TopicPartition, List<Cons
+            RuntimeException exception = null;
+            for (final ConsumerRecord<byte[], byte[]> receivedRecord : receivedRecordsForPartition) {
+                if (!expectedRecord.hasNext())
+                    exception = new RuntimeException("Result verification failed for " + receivedRecord + " since there's no more expected record");
```
Should be `break` for this case?

guozhangwang replied:
It is failed in fact: we expect the received records is no smaller than the expected record.

mjsax replied:
Sure. The question is, why do we not break the `for` loop? If we run out of` expectedRecord` there does not seem to be a reason to iterator to the end of `receivedRecord` (and assign `exception` over and over again)?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/tests/EosTestDriver.java:
```diff
@@ -390,9 +399,12 @@ private static void verifyReceivedAllRecords(final Map<TopicPartition, List<Cons
 
                 if (!receivedKey.equals(expectedKey) || receivedValue != expectedValue) {
-                    throw new RuntimeException("Result verification failed for " + receivedRecord + " expected <" + expectedKey + "," + expectedValue + "> but was <" + receivedKey + "," + receivedValue + ">");
+                    exception = new RuntimeException("Result verification failed for " + receivedRecord + " expected <" + expectedKey + "," + expectedValue + "> but was <" + receivedKey + "," + receivedValue + ">");
```
Should be `break` here?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -461,6 +463,42 @@ public void flush() {
+                            firstException = exception;
+                        } else {
+                            firstException = new ProcessorStateException(
+                                    format("%sFailed to flush cache of store %s", logPrefix, store.name()), exception);
```
nit: formatting:
```
firstException = new ProcessorStateException(
    format("%sFailed to flush cache of store %s", logPrefix, store.name()),
    exception
);
```

guozhangwang replied:
ack

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:
```diff
@@ -38,13 +41,39 @@
+                                    final Map<TopicPartition, Long> newOffsetSnapshot) {
+        // we should always have the old snapshot post completing the register state stores;
+        // if it is null it means the registration is not done and hence we should not overwrite the checkpoint
+        if (oldOffsetSnapshot == null)
```
nit: add `{ }`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:
```diff
@@ -38,13 +41,39 @@
+        // if the previous snapshot is empty while the current snapshot is not then we should always checkpoint;
+        // note if the task is stateless or stateful but no stores logged, the snapshot would also be empty
+        // and hence it's okay to not checkpoint
+        if (oldOffsetSnapshot.isEmpty() && !newOffsetSnapshot.isEmpty())
```
nit: add `{ }`

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateManagerUtil.java:
```diff
@@ -38,13 +41,39 @@
+        if (oldOffsetSnapshot == null)
+            return false;
+
+        // if the previous snapshot is empty while the current snapshot is not then we should always checkpoint;
```
This line just says what the code say itself. The question seem to be, "why" and this question is not answered in the comment.

From my understanding, because we delete the checkpoint file when the task is initialized, we want to write the checkpoint file again as soon as possible (ie, on first commit) instead of waiting until we hit the `OFFSET_DELTA_THRESHOLD_FOR_CHECKPOINT` to keep the time frame in which we have consistent state but not checkpoint file low?

Thus, I am wondering if we should just not delete the checkpoint file when we init the task for non-eos instead? And just remove this additional condition?

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -344,40 +343,41 @@ public void resume() {
 
     /**
+     * @throws StreamsException fatal error that should cause the thread to die
+     * @throws TaskMigratedException recoverable error that would cause the task to be removed
```
It seems like almost every method might throw `StreamsException` and/or `TaskMigratedExcetpion` -- is it really worth to have those comments all over the place?

guozhangwang replied:
I think it is a good habit keeping that since those functions are called by various callers and if the caller do not catch them, the the ancestor callers should -- I've personally found such debugging very helpful.

mjsax replied:
Works for me -- I am just "worried" that those comments might get out-of-sync with the code at some point leading to bug, as we rely on the comments when changing the caller code...

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -344,40 +343,41 @@ public void resume() {
      */
     @Override
     public Map<TopicPartition, OffsetAndMetadata> prepareCommit() {
-        final Map<TopicPartition, OffsetAndMetadata> offsetsToCommit;
```
Why do we remove this? It might be personal preference, but the old code was "cleaner" IMHO? (Feel free to ignore  this comment.)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -428,53 +428,28 @@ public void resume() {
                 // We should never write a checkpoint for a CREATED task as we may overwrite an existing checkpoint
                 // with empty uninitialized offsets
-                log.debug("Skipped writing checkpoint for created task");
+                log.debug("Skipped writing checkpoint for {} task", state());
```
nit: Why this? We know that we are in state `created` ?

guozhangwang replied:
I know, but I just want to make the text consistent (in other places it would use `state()` which is capitalized..) really nit thing.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -428,53 +428,28 @@ public void resume() {
             case RESTORING:
-                writeCheckpoint();
-                log.debug("Finalized commit for restoring task");
+            case SUSPENDED:
```
Nit: Can we keep `SUSPENDED` after `RUNNING` case? We use the same order in all methods and always follow the "natural" state transition order, that is CREATE, RESTORING, RUNNING, SUSPENDED, CLOSED.

guozhangwang replied:
Seems IDEA liked it the other way (it auto merges these two cases).

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -248,12 +248,11 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
-        final Set<Task> dirtyTasks = new HashSet<>();
+        final List<Task> tasksToRecycle = new LinkedList<>();
+        final List<Task> tasksToCloseClean = new LinkedList<>();
+        final List<Task> tasksToCloseDirty = new LinkedList<>();
```
Why do we maintain `List` instead of `Set` ? It seem more natural to me to have a `Set` (ie, should we instead switch `tasksToRecyle` to use a `Set`, too)?

guozhangwang replied:
Two reasons: 1) in unit tests we are validating some scenarios with ordering assumptions, 2) I'd like to have log4j entries to have ordering as well.

ableegoldman replied:
Can we use an ordered set then? Just to make sure we can't end up with a task appearing more than once in the same list/set

guozhangwang replied:
Sounds good, I can try that.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -267,80 +266,19 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+        handleCloseAndRecycle(tasksToRecycle, tasksToCloseClean, tasksToCloseDirty, activeTasksToCreate, standbyTasksToCreate, taskCloseExceptions);
 
         if (!taskCloseExceptions.isEmpty()) {
+            log.error("Hit exceptions while closing / recycling tasks: {}", taskCloseExceptions);
```
Seems like double logging? We have a `log.error` each time before `taskCloseExceptions.put()` is called in `handleCloseAndRecycle`

guozhangwang replied:
Sounds good.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +306,102 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+            } catch (final RuntimeException e) {
+                final String uncleanMessage = String.format("Failed to recycle task %s cleanly. Attempting to close remaining tasks before re-throwing:", task.id());
+                log.error(uncleanMessage, e);
+                taskCloseExceptions.put(task.id(), e);
```
should it be `putIfAbsent` (cf `cleanUpTaskProducer()`)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +306,102 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                        "Failed to close task %s cleanly. Attempting to close remaining tasks before re-throwing:",
+                        task.id());
+                log.error(uncleanMessage, e);
+                taskCloseExceptions.put(task.id(), e);
```
should it be `putIfAbsent` (cf `cleanUpTaskProducer()`)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +306,102 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                        "Failed to checkpoint task %s. Attempting to close remaining tasks before re-throwing:",
+                        task.id());
+                log.error(uncleanMessage, e);
+                taskCloseExceptions.put(task.id(), e);
```
should it be `putIfAbsent` (cf `cleanUpTaskProducer()`)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +306,102 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
 
+        for (final Task task : tasksToCloseClean) {
+            try {
+                if (!tasksToCloseDirty.contains(task)) {
```
This seems to be a "hack" -- IMHO, as task should be only in either one set/list, but never in both... Can we change the first loop to use an explicit iterator and remove a task from `tasksToCloseClean` when we add it to `tasksToCloseDirty`

guozhangwang replied:
I've thought about that, but the tricky thing is that I'm iterating the union of two lists. I can, remove the tasksToCloseDirty from the other two after the iteration to make it cleaner.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +306,102 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                                       final LinkedHashMap<TaskId, RuntimeException> taskCloseExceptions) {
+        if (!tasksToCloseDirty.isEmpty()) {
+            throw new IllegalArgumentException("Tasks to close-dirty should be empty");
+        }
```
While I usually prefer checks like this, it seems unnecessary here? (Feel free to ignore.)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +306,102 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                                    "offsets {} to commit; it means it fails during last commit and hence should be closed dirty",
+                            task.id(), offsets);
+
+                    tasksToCloseDirty.add(task);
```
We should remove `task` from `taskToCloseClean`/`taskToRecycle` here to maintain the invariant that a task is only in either one of the three sets/list but never in multiple at the same time. (cf. my comment below)

I understand that you want to share this loop for "cleanClose" and "recycle" we can extract it into its own method to achieve this.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +306,102 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                taskCloseExceptions.put(task.id(), e);
+                // We've already recorded the exception (which is the point of clean).
+                // Now, we should go ahead and complete the close because a half-closed task is no good to anyone.
+                tasksToCloseDirty.add(task);
```
as above.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +306,102 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+                        task.id());
+                log.error(uncleanMessage, e);
+                taskCloseExceptions.put(task.id(), e);
+                tasksToCloseDirty.add(task);
```
as above

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -368,7 +306,102 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
+        for (final Task task : tasksToRecycle) {
+            final Task newTask;
+            try {
+                if (!tasksToCloseDirty.contains(task)) {
```
as above.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -479,24 +512,20 @@ boolean tryToCompleteRestoration() {
+                remainingRevokedPartitions.removeAll(task.inputPartitions());
             } else if (task.commitNeeded()) {
-                additionalTasksForCommitting.add(task);
+                nonRevokedActiveTasks.add(task);
```
Why this renaming? -- the set does not contain "all" non-revoked active tasks -- only those, that need committing?

guozhangwang replied:
Yeah I think I agree with you -- after a second thought I think this renaming is not very accurate. Will call it `commitNeededActiveTasks`.

mjsax reviewed (commented):
Couple of comments/questions. (Did not look into the tests yet...) Overall, the PR makes sense to me.

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

guozhangwang committed:
do not delete upon loading without eos

guozhangwang commented:
I've addressed the comments and here are the major changes:

1) do not delete checkpoint file upon loading it, and as a result simplify the checkpointNeeded logic, initializing the snapshotLastFlush to the loaded offsets.

2) In closing, also follow the commit -> suspend -> close ordering as in revocation / assignment.

3) If `enforceCheckpoint == true` during RUNNING, still calls maybeCheckpoint even with EOS since that is the case for suspending / closing.

Newly triggered system tests are in https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4094/

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -49,6 +59,30 @@
 
+    protected void initializeCheckpoint() {
+        // we will delete the local checkpoint file after registering the state stores and loading them into the
+        // state manager, therefore we should initialize the snapshot as empty to indicate over-write checkpoint needed
```
Seems the comment is outdated? `we should initialize the snapshot as empty`

guozhangwang replied:
ack.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -93,9 +90,7 @@ public boolean isActive() {
-
-            // initialize the snapshot with the current offsets as we don't need to commit then until they change
-            offsetSnapshotSinceLastCommit = new HashMap<>(stateMgr.changelogOffsets());
+            initializeCheckpoint();
```
In the old code, we actually get a copy of the `Map`, while within `initializeCheckpoint();` don't -- is this on purpose? It it safe?

Also, do we actually need the method? The old code was just doing the exact some thing? It's just one-liner method -- what do we gain?

guozhangwang replied:
Good point, will revert.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -412,7 +412,7 @@ public void shouldInitializeOffsetsFromCheckpointFile() throws IOException {
             stateMgr.initializeStoreOffsetsFromCheckpoint(true);
 
-            assertFalse(checkpointFile.exists());
+            assertTrue(checkpointFile.exists());
```
Should we add a test for EOS, that the checkpoint file is deleted?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -207,15 +207,25 @@ public void shouldFlushAndCheckpointStateManagerOnCommit() {
         task.initializeIfNeeded();
         task.prepareCommit();
-        task.postCommit();
+        task.postCommit(false);  // this should not checkpoint
```
It's unclear to me, how we actually verify that the checkpointing happened? Above, we have
```
stateManager.checkpoint();
EasyMock.expectLastCall();
```

but it only help to verify that we checkpoint a single time, but not which of the three calls does the checkpointing?

guozhangwang replied:
Yeah I tried to mock the static `checkpointNeeded` but it seems not possible with easy-mock.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -1259,13 +1262,47 @@ public void shouldReInitializeTopologyWhenResuming() throws IOException {
+        task.prepareCommit();
+        task.postCommit(false);
+
+        EasyMock.verify(recordCollector);
```
Should we verify `stateManager`, too?

guozhangwang replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -1458,50 +1498,91 @@ public void shouldThrowIfPostCommittingOnIllegalState() {
+                .andReturn(Collections.singletonMap(partition1, 2L))
+                .andReturn(Collections.singletonMap(partition1, 3L));
+        stateManager.checkpoint();
+        EasyMock.expectLastCall().andThrow(new AssertionError("Checkpoint should not be called")).anyTimes();
```
Why do we need to setup an exception? If we don't setup any expected call at all, it should fail automatically?

guozhangwang replied:
Actually if I do not setup, it would not fail.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -1560,42 +1641,13 @@ public void shouldCheckpointWithCreatedStateOnClose() {
-        EasyMock.expectLastCall();
+        EasyMock.expectLastCall().andThrow(new AssertionError("Flush should not be called")).anyTimes();
+        stateManager.checkpoint();
+        EasyMock.expectLastCall().andThrow(new AssertionError("Checkpoint should not be called")).anyTimes();
```
Similar to above: instead of throwing, it should be sufficient to just not register any expected calll?

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -1638,21 +1689,22 @@ public void shouldCheckpointOffsetsOnPostCommitIfCommitNeeded() {
-        stateManager.checkpoint(EasyMock.eq(Collections.singletonMap(partition1, offset)));
-        EasyMock.expectLastCall();
+        stateManager.checkpoint();
+        EasyMock.expectLastCall().andThrow(new AssertionError("Checkpoint should not be called")).anyTimes();
```
as above? (more below... won't add comments each time)

mjsax reviewed (approved):
Just a few follow up. Overall LGTM. Free to merge after addressing (or rejecting) the comments.

guozhangwang committed:
Merge branch 'trunk' of https://github.com/apache/kafka into K9450-decouple-flush-state-from-commit

guozhangwang committed:
github comments
